pkg/store/store.go

```diff
@@ -104,38 +104,51 @@ func (store *OpenTSDBStore) Series(
 func (store *OpenTSDBStore) LabelNames(
 	ctx context.Context,
 	req *storepb.LabelNamesRequest) (*storepb.LabelNamesResponse, error) {
-	labelNames, err := store.openTSDBClient.Suggest(
+	labelNames, err := store.suggestAsList(ctx, "tagk")
+	if err != nil {
+		return nil, err
+	}
+	return &storepb.LabelNamesResponse{
+		Names: labelNames,
+	}, nil
+}
+
+func (store *OpenTSDBStore) suggestAsList(ctx context.Context, t string) ([]string, error) {
+	result, err := store.openTSDBClient.Suggest(
 		opentsdb.SuggestParam{
-			Type:         "tagk",
+			Type:         t,
 			Q:            "",
 			MaxResultNum: math.MaxInt32,
 		})
 	if err != nil {
 		return nil, err
 	}
-	return &storepb.LabelNamesResponse{
-		Names: labelNames.ResultInfo,
-	}, nil
+	return result.ResultInfo, nil
 }
 
 func (store *OpenTSDBStore) LabelValues(
 	ctx context.Context,
 	req *storepb.LabelValuesRequest) (*storepb.LabelValuesResponse, error) {
 	level.Debug(store.logger).Log("msg", "LabelValues", "Label", req.Label)
+	if req.Label == "__name__" {
+		var pNames []string
+		for _, item := range store.metricNames {
+			pNames = append(pNames, strings.Replace(item, ".", ":", -1))
```

Comment:
it would be worth putting it behind a feature flag since this is very CPU and memory-intensive.

Reply:
Makes sense, done.

Reply:
why not just store the list of metric names twice? and then we don't have to do this replacement all the time but just one off every time we load the metrics?

Reply:
ah i see, you're loading the metric names every time for this, that feels.. unnecessary.. excessive?

Reply:
It's not actually loading it, just copying from somewhere in memory. If I was going to optimise this then actually overriding Marshal on the protobuf and caching there would be the way to go (see how https://github.com/grpc/grpc-go/blob/master/encoding/proto/proto.go#L68 allows you to have your own proto.Marshaler), anything else feels a bit pointless as it'd waste memory but still not be optimal... I'll file an issue to look into optimising it later.

Reply:
ah yes, i see now the code has overtaken my comment.